Ticket #35831 Include ModelForm and its Meta options in reference docs (#18699)
django/django: docs/ref/forms/models.txt

> ===========
> ModelForms
> ===========

Reviewer: ```suggestion
==========
ModelForms
==========
```

.. module:: django.forms.models
    :synopsis: ModelForm API reference for inner ``Meta`` class and factory
        functions

.. class:: ModelForm

.. _ref-modelform:

.. currentmodule:: django.forms

:class:`~django.forms.ModelForm` API reference. For introductory material about
ModelForms, see the :doc:`/topics/forms/modelforms` topic guide.

.. _ref-modelform-meta-options:

ModelForm ``Meta`` Options
==========================

The :class:`~django.forms.ModelForm` helper class can be used to build forms
that closely reflect specific Django models. The exact structure of the form
produced can be controlled by setting metadata options as attributes of an inner
``Meta`` class.

> ::

Reviewer 2: Putting `::` on its own line doesn't follow existing style. Suggestion: `Here is an example::`

> ModelForm ``Meta`` Options
> ==========================
> 
> The :class:`~django.forms.ModelForm` helper class can be used to build forms
> that closely reflect specific Django models. The exact structure of the form
> produced can be controlled by setting metadata options as attributes of an inner
> ``Meta`` class.
> 
> ::
> 
>     # Example using some sort of Book model

Reviewer: I was having a look at the [Model _meta API docs](https://docs.djangoproject.com/en/5.1/ref/models/meta/) and was wondering if this should also mention the `_meta` attribute
This would be a suggestion
```suggestion
ModelForm ``_meta`` API
=======================

.. class:: ModelFormOptions

The ``_meta`` API is used to build forms that reflect a Django model. The API
is accessible through the ``_meta`` attribute of each model form, which is an
instance of a ``django.forms.models.ModelFormOptions`` object.

The exact structure of the form produced can be controlled by setting metadata
options as attributes of an inner ``Meta`` class::
```

Reviewer 3: I no longer think that its practical to leave out mention of `ModelFormOptions`, especially because of how the attributes must be referenced in the ref section.

>     # Example using some sort of Book model

Reviewer 2: Try to put text like this outside of comments because it doesn't have an opportunity to be translated.

    from django.forms import ModelForm
    from myapp.models import Book

    class BookForm(ModelForm):
        class Meta:
            model = Book
            fields = ["title", "author"]
            help_texts = {
                "title": "The title of the book",
                "author": "The author of the book",
            }
            # ... other attributes

> .. admonition:: Invalid Meta class attributes

Reviewer: Thinking of the flow, I think this can be above the specific attributes but further down (so above `error_messages`). This information is like an aside so I think can come later

    While any number of attributes can be included in a ``Meta``
    class, **only attributes that match those described below will be used**.
    When troubleshooting or encountering unexpected behavior, be sure to check
    for typos or misnamed attributes because these will **not**, in and of
    themselves, raise an error (though they might cause another error to be
    raised if they affect a required attribute).

> :attr:`~ModelFormOptions.model` is required as well as either 
> :attr:`~ModelFormOptions.fields` or :attr:`~ModelFormOptions.exclude`.
> All other ``Meta`` attributes are optional. Included fields need not be 
> included in every option used. If an option isn't set, or if a model's 
> field isn't used with a particular option, Django will fall back on whatever 
> has already been defined for a field in their model or whatever defaults 
> might exist for that field's type.
> 
> Optional ``Meta`` attributes, other than
> :attr:`~ModelFormOptions.localized_fields` and
> :attr:`~ModelFormOptions.formfield_callback`, expect a dictionary that maps a
> model field name string to some other object or string. Including invalid or
> excluded field name strings in these dictionaries is possible but should have no
> effect since fields that haven't been included would never be accessed.

Reviewer: There's bits in the following suggestion you might not like all of it
```suggestion
Required attributes are :attr:`~ModelFormOptions.model`, and either
:attr:`~ModelFormOptions.fields` or :attr:`~ModelFormOptions.exclude`. All
other ``Meta`` attributes are optional.

Optional attributes, other than :attr:`~ModelFormOptions.localized_fields` and
:attr:`~ModelFormOptions.formfield_callback`, expect a dictionary that maps a
model field name to a value. Any field that is not defined in the dictionary
falls back to the field's default value.
```
I also thought maybe we pull out the invalid field names as another admonition
```
.. admonition:: Invalid field names

    An invalid or excluded field name strings in an optional dictionary
    attribute will have no effect since fields that are not included are not
    accessed.
```

``error_messages``
------------------

.. attribute:: ModelFormOptions.error_messages

    Expects a dictionary that maps a model field name string to a dictionary of
    error message keys (``null``, ``blank``, ``invalid``, ``unique``, etc.)
    mapped to custom error messages.  
    
    Where a field doesn't have an error message dictionary mapped to it in
    :attr:`~ModelFormOptions.error_messages`, Django will fall back on error
    messages defined in that model field's
    :attr:`django.db.models.Field.error_messages` and then finally on the
    default error messages for that field type.

``exclude``
-----------

.. attribute:: ModelFormOptions.exclude

    *One of* :attr:`~ModelFormOptions.fields` *and*

>     :attr:`~ModelFormOptions.exclude` *must be set.* Expects a tuple of strings

Reviewer: Note that a single * is in italics. Not sure if that's on purpose?
I also feel like we can say this last and first say what it is :thinking:

    given :class:`django.db.models.Field` type.

    The :class:`~django.forms.Field` specified in
    :attr:`~ModelFormOptions.field_classes` (the 'new field') gets
    passed all the arguments that would otherwise have been used for that
    model's field (the 'old field'). If the old field has different attributes
    (and therefore different arguments) than the new field, a :exc:`TypeError`
    may be raised. For example:
    
    * Replacing a :class:`~django.forms.CharField` with an
      :class:`~django.forms.EmailField` will work, since the latter is a
      subclass of the former and can handle all the arguments passed to it.
    * Replacing a :class:`~django.forms.CharField` with an
      :class:`~django.forms.IntegerField`, however, will **not** work, since the
      :class:`~django.forms.CharField` would pass a
      :attr:`~django.forms.CharField.max_length` argument, and the
      :class:`~django.forms.IntegerField` cannot accept such an argument (it
      instead has :attr:`~django.forms.IntegerField.max_value`). In this case, a
      :exc:`TypeError` will be raised with the message: 
    
        .. code-block:: console

>             >>> TypeError: Field.__init__() got an unexpected keyword argument 'max_length'

Reviewer 2: This isn't something meant to be typed on the console as `>>>` indicates.

    string ``'__all__'``. Every field named in :attr:`~ModelFormOptions.fields`
    will be included in the output form. Non-editable fields cannot be included.
    
    * If set to the special string ``'__all__'``, all non-editable model fields
      will be included.
    * Other than the special case of ``'__all__'``,
      :attr:`~ModelFormOptions.fields` **cannot** be set to a single string,
      even if only a single field is desired. The input for
      :attr:`~ModelFormOptions.fields` must still be structured as a tuple of
      strings (e.g., ``('myFieldName',)``) or list of strings (e.g., ``[
      'myFieldName' ]``) or else a :exc:`TypeError` will be raised.
    * If a string provided is not a valid field name for the model specified by
      :attr:`ModelFormOptions.model` (e.g., the field doesn't exist on the model
      or the field name contained a typo), a
      :exc:`~django.core.exceptions.FieldError` exception will be raised.
    * If a non-editable field name is provided, a
      :exc:`~django.core.exceptions.FieldError` exception will be raised.
    * If neither of :attr:`~ModelFormOptions.fields` or

>       :attr:`~ModelFormOptions.exclude` are set, an

Reviewer 3: Just an idea to make this more concise: We could specify the valid cases only and mention that the other cases will make Django raise exceptions. The valid cases being either the special string for all fields, and a list or tuple of strings of valid field names, excluding non-editable and fields included in editable.

    though the name of the field can be accessed in the called function using
    ``modelField.name``) and returns a :class:`django.forms.Field` object.

    If :attr:`~ModelFormOptions.formfield_callback` is neither a callable nor a
    function, a :exc:`TypeError` will be raised. 

``help_texts``
--------------

.. attribute:: ModelFormOptions.help_texts

    Expects a dictionary that maps a model field name string to a help text
    string. 
    
    Where a field doesn't have a label string mapped to it in
    :attr:`~ModelFormOptions.help_texts`, Django will fall back on that model
    field's :attr:`~django.db.models.Field.help_text`.

``labels``
----------

.. attribute:: ModelFormOptions.labels

    Expects a dictionary that maps a model field name string to a label string. 
    
    Where a field doesn't have a label string mapped to it in
    :attr:`~ModelFormOptions.labels`, Django will fall back on that model
    field's :attr:`~django.db.models.Field.verbose_name` and then the field's
    attribute name.

``localized_fields``
--------------------

.. attribute:: ModelFormOptions.localized_fields

>     Expects a tuple of strings or a list of strings that are field names.

Reviewer 3: `A tuple or list of strings representing field names.`

    
    The fields included in this list will have their data localized (by default,
    the form fields output by a :class:`~django.forms.ModelForm` are **not**
    localized, see :ref:`Creating forms from models: Enabling localization of
    fields <modelforms-enabling-localization-of-fields>`).  
    
    Like :attr:`~ModelFormOptions.fields`,
    :attr:`~ModelFormOptions.localized_fields` can be set to the special string
    ``'__all__'`` to include all fields.

``model``
---------

.. attribute:: ModelFormOptions.model

    *Required.* A :class:`django.db.models.Model` to be used for this
    :class:`~django.forms.ModelForm`.

    If no class is specified, a :exc:`ValueError` exception will be raised.

``widgets``
-----------

.. attribute:: ModelFormOptions.widgets

    Expects a dictionary that maps a model field name string to a
    :class:`django.forms.Widget`. 
    
    Where an included model field doesn't have a widget mapped to it in
    :attr:`~ModelFormOptions.widgets`, Django will fall back on the default
    widget for that particular type of :class:`django.db.models.Field`.

> ModelForm Factory Functions

Reviewer 2: As above, only the first word should be capitalized.

===========
ModelForms
===========

.. module:: django.forms.models
    :synopsis: ModelForm API reference for inner ``Meta`` class and factory
        functions

.. class:: ModelForm

.. _ref-modelform:

.. currentmodule:: django.forms

:class:`~django.forms.ModelForm` API reference. For introductory material about
ModelForms, see the :doc:`/topics/forms/modelforms` topic guide.

.. _ref-modelform-meta-options:

ModelForm ``Meta`` Options
==========================

The :class:`~django.forms.ModelForm` helper class can be used to build forms
that closely reflect specific Django models. The exact structure of the form
produced can be controlled by setting metadata options as attributes of an inner
``Meta`` class.

::

    # Example using some sort of Book model

    from django.forms import ModelForm

>     from myLibraryApp.models import Book

Reviewer 4: Nitpick comment:

```suggestion
    from myapp.models import Book
```
It looks like the majority of code examples would use `myapp`.

Author: Good point-- I should be consistent with that.

As a sidebar, is there a convention anywhere in the docs for contributors about what can/should be used as an example (where possible) for consistency across the docs?

Reviewer 4: Ya know I checked the codebase and I didn't see perfect uniformity. However, `myapp` was by far the most common. I probably wouldn't have said something if it were using snake case either.

    class BookForm(ModelForm):
        class Meta:
            model = Book
            fields = ["title", "author"]
            help_texts = {
                "title": "The title of the book",
                "author": "The author of the book",
            }
            # ... other attributes

.. admonition:: Invalid Meta class attributes

    While any number of attributes can be included in a ``Meta``
    class, **only attributes that match those described below will be used**.
    When troubleshooting or encountering unexpected behavior, be sure to check
    for typos or misnamed attributes because these will **not**, in and of
    themselves, raise an error (though they might cause another error to be
    raised if they affect a required attribute).

> Other than :attr:`~ModelFormOptions.model` and at least one of
> :attr:`~ModelFormOptions.fields` and :attr:`~ModelFormOptions.exclude`, ``Meta``
> attributes are optional. It is **not** necessary for every option to be set for

Reviewer 4: I think this can be improved if we list things in a more definitive manner such as:

```suggestion
:attr:`~ModelFormOptions.model` is required as well as either 
:attr:`~ModelFormOptions.fields` or :attr:`~ModelFormOptions.exclude`.
All other ``Meta`` attributes are optional.
```

I think this is clearer because it indicates the required state and the fields immediately rather than listing things, then indicating that this list is required, while everything else is optional.

I don't think this is a required change.

each ``Meta`` class, nor does every included field need to be included in every
option used. If an option isn't set, or if a model's field isn't used with a
particular option, Django will fall back on whatever has already been defined
for a field in their model or whatever defaults might exist for that field's
type.

Optional ``Meta`` attributes, other than
:attr:`~ModelFormOptions.localized_fields` and
:attr:`~ModelFormOptions.formfield_callback`, expect a dictionary that maps a
model field name string to some other object or string. Including invalid or
excluded field name strings in these dictionaries is possible but should have no
effect since fields that haven't been included would never be accessed.

``error_messages``
------------------

.. attribute:: ModelFormOptions.error_messages

>     Expects a dictionary that maps a model field name string to a dictionary of

Reviewer 4: I noticed that there are some other areas of the documentation that are similar, but use different language. Specifically in `ModelAdmin`'s reference docs (https://django--18699.org.readthedocs.build/en/18699/ref/contrib/admin/index.html#django.contrib.admin.ModelAdmin.prepopulated_fields). It looks like most places uses something like "Set <field_name> to ..."

Following that same pattern may make translations easier down the line because it'll be similar phrasing. It should require less consideration of what the English phrase means since there are other cases of it.

This is a minor thing I noticed, not sure if it needs to change though.

Reviewer: ```suggestion
    A dictionary that maps a model field name to a dictionary of error
```
True, my personal preference is it just stating what it is :grin:

================================================

> ModelForm ``Meta`` Options and Factory Functions

Reviewer 3: ```suggestion
Model Forms
```

================================================

.. module:: django.forms.models
    :synopsis: ModelForm API reference for inner ``Meta`` class and factory
        functions

.. admonition:: Other ModelForm-related documentation

    This document is an API reference for the :class:`~django.forms.ModelForm`
    class. For examples or additional context, please refer to the
    documentation topic :doc:`Creating forms from models

>     </topics/forms/modelforms>`.

Reviewer 3: I believe the previous note was sufficient. There's no need for an admonition here.

```
Model Form API reference. For introductory material about model forms, see the
:doc:`/topics/forms/modelforms` topic guide.
```

.. _ref-modelform-meta-options:

> ModelForm ``Meta`` Options

Reviewer 3: Maybe first add a reference for `ModelForm` itself.

    class BookForm(ModelForm)
        class Meta:
            model = Book
            fields = [ 'title', 'author' ]
            help_texts = {
                'title' : 'The title of the book',
                'author': 'The author of the book'
            }
            # ... other attributes

.. admonition:: Invalid Meta class attributes

    While any number of attributes can be included in a ``Meta``
    class, **only attributes that match those described below will be used**.
    When troubleshooting or encountering unexpected behavior, be sure to check
    for typos or misnamed attributes because these will **not**, in and of
    themselves, raise an error (though they might cause another error to be
    raised if they affect a required attribute).

It is **not** necessary for every option to be set for each ``Meta`` class,
though some (indicated below) are required.

In most cases, if an option isn't set, or if a model's field isn't used with a
particular option, Django will fallback on whatever has already been defined for
a field in their model or whatever defaults might exist for that field's type.

.. _ref-modelform-meta-required:

> Required ``Meta`` Options

Reviewer 3: No need to group the attributes this way, since attributes specify whether they're optional or required.

Author: Yep-- I've reordered them to be purely alphabetically now.

----------------------------

``model``
~~~~~~~~~

.. attribute:: ModelFormOptions.model

    *Required.* A :class:`django.db.models.Model` to be used for this

>     :class:`~django.forms.ModelForm`. This is **not** a string.

Reviewer 3: I'd remove the note "this is not a string"
